### PR TITLE
[4.0] Dependency and build script tweaks

### PR DESCRIPTION
**Pull request by C-Lodder**
Pull Request for Issue # .

### Summary of Changes

1. Remove `uglifycss` and use postcss/cssnano like we do in other areas of the build scripts
2. Remove the promised based `recursive-readdir` dependency and use a small asynchronous custom function to recursively read directories.
3. Update NodeJS + NPM dependencies to latest LTS
4. Update some dev dependencies to their respective latest version

Still room for improvements but this is a start

### Testing Instructions

1. Run `npm i`
2. Ensure there are no errors in the terminal when building

@dgrammatiko thoughts on this?

### CODE REVIEW

**Comment by dgrammatiko** [comment]
@C-Lodder it's fine I guess, just a reminder that last time I tried to use async/await George was against it, thus the promise based is still here

File: .eslintrc [hunk 1]
```diff
@@ -9,6 +9,11 @@
     "es6": true,
     "node": true
   },
+  "parserOptions": {
+    // Enable parsing of asyc functions
+    "ecmaVersion": 2018
+  },
+  },
```
Comment (wilsonge) [inline .eslintrc]: syntax

**Comment by dgrammatiko** [comment]
Also something else since you're touching this:
in the comile-scss change this [~line 44]:
```js
        files = [
          `${RootPath}/templates/cassiopeia/scss/offline.scss`,
          `${RootPath}/templates/cassiopeia/scss/template.scss`,
          `${RootPath}/templates/cassiopeia/scss/template-rtl.scss`,
          `${RootPath}/administrator/templates/atum/scss/template.scss`,
          `${RootPath}/administrator/templates/atum/scss/template-rtl.scss`,
          `${RootPath}/administrator/templates/atum/scss/system/searchtools/searchtools.scss`,
          `${RootPath}/administrator/templates/atum/scss/vendor/awesomplete/awesomplete.scss`,
          `${RootPath}/administrator/templates/atum/scss/vendor/choicesjs/choices.scss`,
          `${RootPath}/administrator/templates/atum/scss/vendor/minicolors/minicolors.scss`,
          `${RootPath}/administrator/templates/atum/scss/vendor/joomla-custom-elements/joomla-alert.scss`,
          `${RootPath}/administrator/templates/atum/scss/vendor/joomla-custom-elements/joomla-tab.scss`,
          `${RootPath}/administrator/templates/atum/scss/vendor/fontawesome-free/fontawesome.scss`,
          `${RootPath}/installation/template/scss/template.scss`,
          `${RootPath}/installation/template/scss/template-rtl.scss`,
        ];
        folders = [
          `${RootPath}/build/media_source`,
        ];
```
to 

```js
        files = [
          `${RootPath}/installation/template/scss/template.scss`,
          `${RootPath}/installation/template/scss/template-rtl.scss`,
        ];
        folders = [
          `${RootPath}/build/media_source`,
`${RootPath}/administrator/templates`,
`${RootPath}/templates`,
        ];
```

That way someone can use these tools to build any front end or back end template

**Comment by dgrammatiko** [comment]
Also something else since you're touching this:
in the comile-scss change this [~line 44]:
```js
        files = [
          `${RootPath}/templates/cassiopeia/scss/offline.scss`,
          `${RootPath}/templates/cassiopeia/scss/template.scss`,
          `${RootPath}/templates/cassiopeia/scss/template-rtl.scss`,
          `${RootPath}/administrator/templates/atum/scss/template.scss`,
          `${RootPath}/administrator/templates/atum/scss/template-rtl.scss`,
          `${RootPath}/administrator/templates/atum/scss/system/searchtools/searchtools.scss`,
          `${RootPath}/administrator/templates/atum/scss/vendor/awesomplete/awesomplete.scss`,
          `${RootPath}/administrator/templates/atum/scss/vendor/choicesjs/choices.scss`,
          `${RootPath}/administrator/templates/atum/scss/vendor/minicolors/minicolors.scss`,
          `${RootPath}/administrator/templates/atum/scss/vendor/joomla-custom-elements/joomla-alert.scss`,
          `${RootPath}/administrator/templates/atum/scss/vendor/joomla-custom-elements/joomla-tab.scss`,
          `${RootPath}/administrator/templates/atum/scss/vendor/fontawesome-free/fontawesome.scss`,
          `${RootPath}/installation/template/scss/template.scss`,
          `${RootPath}/installation/template/scss/template-rtl.scss`,
        ];
        folders = [
          `${RootPath}/build/media_source`,
        ];
```
to 

```js
        files = [
          `${RootPath}/installation/template/scss/template.scss`,
          `${RootPath}/installation/template/scss/template-rtl.scss`,
        ];
        folders = [
          `${RootPath}/build/media_source`,
`${RootPath}/administrator/templates`,
`${RootPath}/templates`,
        ];
```

That way someone can use these tools to build any front end or back end template

**Comment by C-Lodder** [comment]
@dgrammatiko Template devs should be using their own build environment rather than being tied in to Joomla's. Also, if we add this support in, it will actually prevent template devs using their own. Just tested with my template and it failed

@wilsonge I'm ignoring Hound errors till `package.json` and Hound are using the the latest version of eslint, they're matching, and the parser options are set properly

File: package.json [hunk 2]
```diff
@@ -8,8 +8,8 @@
     "url": "https://github.com/joomla/joomla-cms.git"
   },
   "engines": {
-    "node": ">=8.11.4",
-    "npm": ">=5.7.0"
+    "node": ">=12.16.1",
```
Comment (mbabker) [inline package.json:11]: Jenkins needs to be updated before merging this (right now it's on the 10.x LTS).  Otherwise nightly builds will fail.

**Comment by dgrammatiko** [comment]
> Just tested with my template and it failed

Are you using the same conventions? eg src= `scss` dest: `css`?


>  it will actually prevent template devs using their own

No it won't! You can still `npm init` any subfolder in the repo

**Comment by wilsonge** [comment]
> @wilsonge I'm ignoring Hound errors till package.json and Hound are using the the latest version of eslint, they're matching, and the parser options are set properly

Then the PR's probably going to sit here a long time. I'm not going to go run around do this for you - this is nice to have but no where close to a priority.

**Comment by C-Lodder** [comment]
Not expecting to you do anything. Will leave this here till it's fixed.

**Pull request closed without merging**